rak/functional.h

```diff
@@ -1,183 +1,18 @@
-struct reference_fix {
-  typedef Type type;
-};
-
-template <typename Type>
-struct reference_fix<Type&> {
-  typedef Type type;
-};
-
-// Operators:
-
-template <typename Type, typename Ftor>
-struct equal_t {
-  typedef bool result_type;
-
-  equal_t(Type t, Ftor f) : m_t(t), m_f(f) {}
-
-  template <typename Arg>
-  bool operator () (Arg& a) {
-    return m_t == m_f(a);
-  }
-
-  Type m_t;
-  Ftor m_f;
-};
-
-template <typename Type, typename Ftor>
-inline equal_t<Type, Ftor>
-equal(Type t, Ftor f) {
-  return equal_t<Type, Ftor>(t, f);
-}
-
-template <typename Type, typename Ftor>
-struct less_t {
-  typedef bool result_type;
-
-  less_t(Type t, Ftor f) : m_t(t), m_f(f) {}
-
-  template <typename Arg>
-  bool operator () (Arg& a) {
-    return m_t < m_f(a);
-  }
-
-  Type m_t;
-  Ftor m_f;
-};
-
-template <typename Type, typename Ftor>
-inline less_t<Type, Ftor>
-less(Type t, Ftor f) {
-  return less_t<Type, Ftor>(t, f);
-}
-
-template <typename Type, typename Ftor>
-struct less_equal_t {
-  typedef bool result_type;
-
-  less_equal_t(Type t, Ftor f) : m_t(t), m_f(f) {}
-
-  template <typename Arg>
-  bool operator () (Arg& a) {
-    return m_t <= m_f(a);
-  }
-
-  Type m_t;
-  Ftor m_f;
-};
-
-template <typename Type, typename Ftor>
-inline less_equal_t<Type, Ftor>
-less_equal(Type t, Ftor f) {
-  return less_equal_t<Type, Ftor>(t, f);
-}
-
-template <typename Src, typename Dest>
-struct on_t : public std::unary_function<typename Src::argument_type, typename Dest::result_type> {
-  typedef typename Dest::result_type result_type;
-
-  on_t(Src s, Dest d) : m_dest(d), m_src(s) {}
-
-  result_type operator () (typename reference_fix<typename Src::argument_type>::type arg) {
-    return m_dest(m_src(arg));
-  }
-
-  Dest m_dest;
-  Src m_src;
-};
-    
-template <typename Src, typename Dest>
-inline on_t<Src, Dest>
-on(Src s, Dest d) {
-  return on_t<Src, Dest>(s, d);
-}  
-
-template <typename T>
-struct call_delete : public std::unary_function<T*, void> {
-  void operator () (T* t) {
-    delete t;
-  }
-};
-
-template <typename Container>
+template <typename Container, typename... Args>
 inline void
-slot_list_call(const Container& slot_list) {
-  if (slot_list.empty())
-    return;
-
-  typename Container::const_iterator first = slot_list.begin();
-  typename Container::const_iterator next = slot_list.begin();
-
-  while (++next != slot_list.end()) {
-    (*first)();
-    first = next;
-  }
-
-  (*first)();
-}
-
-template <typename Container, typename Arg1>
-inline void
-slot_list_call(const Container& slot_list, Arg1 arg1) {
-  if (slot_list.empty())
-    return;
-
-  typename Container::const_iterator first = slot_list.begin();
-  typename Container::const_iterator next = slot_list.begin();
-
-  while (++next != slot_list.end()) {
-    (*first)(arg1);
-    first = next;
+slot_list_call(const Container& slot_list, Args&&... args) {
+  for (const auto& slot : slot_list) {
+    slot(std::forward<Args>(args)...);
   }
```

Comment:
When calling slot the list entry can remove itself.

This is why the container must be a std::list (or equivalently container type) and we need to have the next iterator entry before calling slot.

Reply:
Gotcha, reverted to original code.